src/reducers/comments.tsx

```diff
@@ -71,14 +71,16 @@ export type CommentsByKey = {
 
 export type CommentsState = {
   byKey: CommentsByKey;
-  byId: {
-    [id: number]: Comment;
-  };
+  byId: { [id: number]: Comment };
+  forVersionId: undefined | number;
+  hasComments: undefined | boolean;
```

Comment:
> Wouldn't `hasComments` be the same as `state.byKey !== {}` 

No, because when the user is entering a comment (before saving it), `state.byKey` looks something like:

```js
{
  'manifest.json': {
    beginNewComment: true,
    commentIds: [],
  }
}
```

I didn't want to loop through each and every `CommentInfo` object to check all lengths of `commentIds`.

> ...and/or `state.byId !== {}`?

It may be possible to just check the length of `state.byId` but it can get tricky to maintain that object over time. We haven't implemented comment deletion yet but we could get into a situation where an ID is removed from a list while its object is not deleted from `byId`.

More importantly, I haven't added all the necessary code to keep `byId` in sync when switching the value of `forVersionId`. Specifically, `byId` does not get reset when switching versions. I fear that the complexity needed for that would outweigh the simplicity of removing the `hasComments` maintenance. It's not totally clear when I'd need to reset `byId`. On every action, I guess?

🤷‍♂ 

Maybe I'll give it a try!

Reply:
I switched to always expunging old values when switching versions.